MiniF/communications/consumers.py

```diff
@@ -57,5 +109,86 @@ async def chat_message(self, event):
         - Extracts the message from the event.
         - Sends the message to the WebSocket as JSON.
         """
-        message = event["message"]
-        await self.send(text_data=json.dumps({"message": message}))
+        await self.send(text_data=json.dumps({"message": event["message"], "sender": event["sender"]}))
+
+    @database_sync_to_async
+    def get_room(self, room_id):
+        try:
+            return ChatRoom.objects.get(id=room_id)
+        except ChatRoom.DoesNotExist:
+            return None
+
+    @database_sync_to_async
+    def is_user_participant(self):
+        if not self.room:
+            return False
+        return self.scope["user"] in [self.room.investor, self.room.startup]
+
+    @database_sync_to_async
+    def get_receiver(self):
+        if not self.room:
+            return None
+        current_user = self.scope["user"]
+        return self.room.startup if current_user == self.room.investor else self.room.investor
+
+    @database_sync_to_async
+    def get_user(self, user_id):
+        """
+        Handles broadcasting messages to WebSocket clients.
+
+        - Sends the message and sender information to the client as JSON.
+        """
+        return User.objects.get(pk=user_id)
+
+    @database_sync_to_async
+    def get_or_create_chat_room(self, user1, user2):
+        """
+        Retrieves or creates a chat room between two users.
+
+        Args:
+            user1 (User): The first user.
+            user2 (User): The second user.
+
+        Returns:
+            ChatRoom: The existing or newly created chat room.
+        """
+        try:
+            room = ChatRoom.objects.get_or_create(
+                investor=user1,
+                startup=user2,
+            )
+            return room[0]
+        except IntegrityError:
+            return ChatRoom.objects.get(
+                investor=user1,
+                startup=user2,
+            )
+
+    @database_sync_to_async
+    def save_message(self, room, sender, receiver, content):
+        """
+        Saves a message to the database.
+
+        Args:
+            room (ChatRoom): The chat room where the message is sent.
+            sender (User): The sender of the message.
+            receiver (User): The receiver of the message.
+            content (str): The message content.
+
+        Returns:
+            Message: The created message object.
+        """
+        try:
+            message = Message(
+                room_id=room.pk,
+                sender_id=sender.pk,
+                receiver_id=receiver.pk,
+                content=content,
+                timestamp=datetime.datetime.now(),
+            )
+            message.save()
+            if not message.id:
+                raise PyMongoError("Message not saved")
+        except PyMongoError as e:
+            print(f"MongoDB error: {e}")
```

Comment:
In except PyMongoError you only do print. In production it is better through logging.
This functionality is ready right now.